simulation-with-simpy/simulate.py

```diff
@@ -0,0 +1,118 @@
+"""Companion code to https://realpython.com/simulation-with-simpy/
+
+'Simulating Real-World Processes With SimPy'
+
+Python version: 3.7.3
+SimPy version: 3.0.11
+"""
+
+import simpy
+import random
+
+arrival_times = []
+departure_times = []
+
+
+class Theater(object):
+    def __init__(self, env, num_cashiers, num_servers, num_ushers):
+        self.env = env
+        self.cashier = simpy.Resource(env, num_cashiers)
+        self.server = simpy.Resource(env, num_servers)
+        self.usher = simpy.Resource(env, num_ushers)
+
+    def purchase_ticket(self, moviegoer):
+        yield self.env.timeout(random.randint(1, 3))
+
+    def check_ticket(self, moviegoer):
+        yield self.env.timeout(3 / 60)
+
+    def sell_food(self, moviegoer):
+        yield self.env.timeout(random.randint(1, 5))
+
+
+def go_to_movies(env, moviegoer, theater):
+    with theater.cashier.request() as request:
+        yield request
+        yield env.process(theater.purchase_ticket(moviegoer))
+
+    with theater.usher.request() as request:
+        yield request
+        yield env.process(theater.check_ticket(moviegoer))
+
+    get_food = [0, 1]
+    if random.choice(get_food) == 1:
+        with theater.server.request() as request:
+            yield request
+            yield env.process(theater.sell_food(moviegoer))
+
+    departure_time = env.now
+    departure_times.append(departure_time)
+
+
+def run_theater(env, num_cashiers, num_servers, num_ushers):
+    theater = Theater(env, num_cashiers, num_servers, num_ushers)
+
+    for moviegoer in range(3):
+        env.process(go_to_movies(env, moviegoer, theater))
+        arrival_time = env.now
+        arrival_times.append(arrival_time)
+
+    while True:
+        yield env.timeout(0.20)  # Wait a bit before generating a new person
+
+        moviegoer += 1
+        env.process(go_to_movies(env, moviegoer, theater))
+        arrival_time = env.now
+        arrival_times.append(arrival_time)
+
+
+def calculate_wait_time(arrival_times, departure_times):
+    total_wait = []
+    total_seated = len(departure_times)
+
+    for i in range(total_seated):
+        total_wait.append(departure_times[i] - arrival_times[i])
```

Comment:
Two different ideas here:

1. since both arrival and departure times are globals, why not replace `departure_times` with `wait_times` and compute each wait time when you know the departure?
2.  If that's not appealing, then this could potentially be a list comprehension using enumerate to get the index from one.

I'd vote for option 1 as it's cleaner. :)

Reply:
Done! In addition, I found a way to have `wait_times` as the only global. Check [this comment](https://github.com/realpython/tutorial-drafts/pull/337#discussion_r315884543) on the tutorial PR for more info.